bundles/org.openhab.binding.entsoe/src/main/java/org/openhab/binding/entsoe/internal/EntsoeHandler.java

```diff
@@ -0,0 +1,280 @@
+        } catch (EntsoeResponseMapException e) {
+            logger.warn("Using millis to next whole hour");
+        }
+        return millis;
+    }
+
+    private Optional<EntsoeTimeSerie> getActualTimeSerie() throws EntsoeResponseMapException {
+        return getActualTimeSerie(false);
+    }
+
+    private Optional<EntsoeTimeSerie> getActualTimeSerie(boolean getNext) throws EntsoeResponseMapException {
+        if (entsoeTimeSeries.isEmpty()) {
+            throw new EntsoeResponseMapException("responseMap is empty");
+        }
+        Duration duration = Duration.parse(config.resolution);
+        for (Map.Entry<Instant, Optional<EntsoeTimeSerie>> entry : entsoeTimeSeries.entrySet()) {
+            Instant thisEntry = entry.getKey();
+            Instant now = Instant.now();
+            Instant nowPlusNext = now.plus(duration);
+            Instant nowMinusNext = now.minus(duration);
+
+            if (getNext && thisEntry.isAfter(now)) {
+                return entry.getValue();
+            }
+            if (!getNext && thisEntry.isAfter(nowMinusNext) && thisEntry.isBefore(nowPlusNext)) {
+                return entry.getValue();
+            }
+        }
+        throw new EntsoeResponseMapException("Could not get actual ENTSOE TimeSerie");
+    }
+
+    private boolean needToFetchHistoricDays() {
+        return needToFetchHistoricDays(false);
+    }
+
+    private boolean needToFetchHistoricDays(boolean updateHistoricDaysInitially) {
+        boolean needToFetch = false;
+        if (historicDaysInitially < config.historicDays) {
+            logger.debug("Need to fetch historic data. Historicdays was changed to a greater number: {}",
+                    config.historicDays);
+            needToFetch = true;
+        }
+
+        if (updateHistoricDaysInitially && historicDaysInitially != config.historicDays) {
+            historicDaysInitially = config.historicDays;
+        }
+
+        return needToFetch;
+    }
+
+    private void refreshPrices() {
+        if (!isLinked(EntsoeBindingConstants.CHANNEL_SPOT_PRICE)) {
+            logger.debug("Channel {} is not linked, skipping channel update",
+                    EntsoeBindingConstants.CHANNEL_SPOT_PRICE);
+            return;
+        }
+
+        config = getConfigAs(EntsoeConfiguration.class);
+
+        if (entsoeTimeSeries.isEmpty()) {
+            fetchNewPrices();
+            return;
+        }
+
+        boolean needsInitialUpdate = lastDayAheadReceived.equals(ZonedDateTime.of(LocalDateTime.MIN, cetZoneId))
+                || needToFetchHistoricDays(true);
+        boolean hasNextDayValue = false;
+        for (Map.Entry<Instant, Optional<EntsoeTimeSerie>> entry : entsoeTimeSeries.entrySet()) {
+            if (entry.getKey().atZone(cetZoneId)
+                    .isAfter(currentCetTimeWholeHours().withHour(config.spotPricesAvailableCetHour))) {
+                hasNextDayValue = true;
+                break;
+            }
+        }
```

Comment:
This could be optimized by instantiating the target time only once outside the loop and eliminating the entry lookup by using `keySet()` rather than `entrySet()`. Perhaps we could also use the Stream API to make it more concise:
```suggestion
        Instant availabilityTime = currentCetTimeWholeHours().withHour(13).toInstant();
        boolean hasNextDayValue = entsoeTimeSeries.keySet().stream()
                .anyMatch(instant -> instant.isAfter(availabilityTime));
```

Reply:
I see you went with a slightly modified version:
```java
Instant nextDayInstant = ZonedDateTime.now(cetZoneId).plusDays(1).with(LocalTime.MIDNIGHT).toInstant();
```

rather than the previous logic rewritten:
```java
Instant availabilityTime = currentCetTimeWholeHours().withHour(13).toInstant();
```

Was this intentional? In this case you wouldn't even need to iterate to check if anything is after midnight, you can just check the key directly:
```java
        Instant nextDayInstant = ZonedDateTime.now(cetZoneId).plusDays(1).with(LocalTime.MIDNIGHT).toInstant();
        boolean hasNextDayValue = entsoeTimeSeries.containsKey(nextDayInstant);
```

Does it make sense?

Reply:
Yes, that should work for all resolutions 15, 30 and 60.